Add default namespace to client

dmcgowan opened the pull request:
Adds a default namespace to the client. Intercepts client requests to add namespace to context.

crosbymichael commented:
Why not just use https://godoc.org/google.golang.org/grpc#WithUnaryInterceptor in the client and modify the context that way?

dmcgowan commented:
Removed the refactoring and updated to use the interceptor pattern. In the future if we want to use multiple interceptors we can make an interceptor chain.

crosbymichael commented on client.go:
```diff
@@ -41,38 +41,55 @@ func init() {
 }
 
-type NewClientOpts func(c *Client) error
+type clientOpts struct {
```
This should probably be exported

dmcgowan replied:
Do we need to export it though? Everything we add here should have a corresponding `With` function, having this unexported prevents callers from implementing the these option functions themselves rather than using the ones already existing ones. Correspondingly everything in this struct must then be exported to be useful.  Most uses of `With*` patterns I have seen use unexported types here to have better control of the configuration (such as grpc, docker distribution)

stevvooe commented:
This looks good but it needs stream support.

dmcgowan commented:
Updated to add stream interceptor as well

crosbymichael commented:
LGTM

stevvooe commented:
LGTM
